pkg/controller/cloudinit/controller.go

```diff
@@ -0,0 +1,231 @@
+package cloudinit
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+
+	ctlnodev1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	cloudinitv1 "github.com/harvester/node-manager/pkg/apis/node.harvesterhci.io/v1beta1"
+	"github.com/harvester/node-manager/pkg/cloudinit"
+	ctrlv1 "github.com/harvester/node-manager/pkg/generated/controllers/node.harvesterhci.io/v1beta1"
+)
+
+const handlerName = "harvester-node-cloud-init-controller"
+
+type controller struct {
+	nodeName   string
+	cloudinits ctrlv1.CloudInitClient
+	nodeCache  ctlnodev1.NodeCache
+}
+
+func Register(ctx context.Context, nodeName string, cloudinits ctrlv1.CloudInitController, nodeCache ctlnodev1.NodeCache) {
+	ctl := &controller{
+		nodeName:   nodeName,
+		cloudinits: cloudinits,
+		nodeCache:  nodeCache,
+	}
+
+	cloudinits.OnChange(ctx, handlerName, ctl.OnCloudInitChange)
+}
+
+func (c *controller) OnCloudInitChange(_ string, cloudInitObj *cloudinitv1.CloudInit) (*cloudinitv1.CloudInit, error) {
+	if cloudInitObj == nil || cloudInitObj.DeletionTimestamp != nil {
+		return cloudInitObj, nil
+	}
+
+	cloudInitCopy := cloudInitObj.DeepCopy()
+
+	if cloudInitCopy.Annotations == nil {
+		cloudInitCopy.Annotations = make(map[string]string)
+	}
+
+	checksum, err := cloudinit.Measure(strings.NewReader(cloudInitCopy.Spec.Contents))
+	_ = err
+
+	checksumString := fmt.Sprintf("%x", checksum)
+	persistedChecksum := strings.ToLower(cloudInitCopy.Annotations[cloudinit.AnnotationHash])
+
+	if checksumString != persistedChecksum {
+		cloudInitCopy.Annotations[cloudinit.AnnotationHash] = checksumString
+		return c.cloudinits.Update(cloudInitCopy)
+	}
+
+	node, err := c.nodeCache.Get(c.nodeName)
+	if err != nil {
+		return cloudInitCopy, err
+	}
+
+	if cloudinit.MatchesNode(node, cloudInitCopy) {
+		_ = cloudinit.RequireLocal(cloudInitCopy)
```

Comment:
Should we need to do the error handling here?
Or do we have another mechanism to ensure the cloudinit is synced with local?

Reply:
Yeah, I think this should call update status and then return the error so it gets requeued